### PR TITLE
Support DEFAULT keyword and '-' prefix in SSL_CTX_set1_groups_list()

### LINKED ISSUE #25790: Configuring available TLS groups using negation
We are setting up some tests that should confirm correct TLS configuration of our application.

We'd like to show that for example the client will fail if a server is not configured to use `x25519` so we want to do something similar to

```
[tls_system_default]
SSLCipherSuite = DEFAULT:!kDHE
```

ie.

```
[tls_system_default]
Groups = DEFAULT:!x25519
```

Is that currently possible or is it feasible as a feature request?

### PR DESCRIPTION
**Pull request by fwh-dc**
Fixes #25790

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/t1_lib.c [hunk 1]
```diff
@@ -1046,24 +1046,71 @@ static int gid_cb(const char *elem, int len, void *arg)
...
+    size_t groups_to_add = 0;
 
     if (elem == NULL)
         return 0;
-    if (elem[0] == '?') {
-        ignore_unknown = 1;
-        ++elem;
-        --len;
+
+    while (elem[0] == '-' || elem[0] == '?') {
```
[the remark this answers is not in the document]
Comment (fwh-dc) [reply ssl/t1_lib.c]: You're right. Fixed.

**Review by mattcaswell — APPROVED** [review]
Looks good. Consider adding a CHANGES.md entry

**Comment by mattcaswell** [comment]
CIs are failing

**Comment by mattcaswell** [comment]
There's still some CI failures with the latest push

**Comment by fwh-dc** [comment]
> There's still some CI failures with the latest push

Yes the issue is with these lines in minimal builds:

https://github.com/openssl/openssl/blob/71cd2a493a9b9570cc99b89199034b04f8a8db52/test/sslapitest.c#L9784-L9786

Is it correct that the length is 0? And which definition can I use for enabling the check?

**Comment by mattcaswell** [comment]
I suppose the problem comes with the combination of no-ec and no-dh. That leaves you with no groups left.

**Comment by mattcaswell** [comment]
i.e. `!defined(OPENSSL_NO_EC) || !defined(OPENSSL_NO_DH))`, or something like that

**Comment by mattcaswell** [comment]
There's quite a few other tests in that file gated on those defines

**Comment by fwh-dc** [comment]
> I suppose the problem comes with the combination of no-ec and no-dh. That leaves you with no groups left.

Thanks! I did not know that was possible. I'm pretty sure the update works.

File: test/sslapitest.c [hunk 2]
```diff
@@ -9782,22 +9782,30 @@ static int test_configuration_of_groups(void)
...
 
     default_groups_len = ctx->ext.supported_groups_default_len;
-
+#if (!defined(OPENSSL_NO_EC) || !defined(OPENSSL_NO_DH))
     if (!TEST_size_t_gt(default_groups_len, 0)
         || !TEST_int_gt(SSL_CTX_set1_groups_list(ctx, "DEFAULT"), 0)
+#else
+    if (!TEST_size_t_eq(default_groups_len, 0)
+        || !TEST_int_eq(SSL_CTX_set1_groups_list(ctx, "DEFAULT"), 0)
+#endif
```
Comment (t8m) [inline test/sslapitest.c:9791]: I would not add these checks as in future they might be misleading, i.e. when we will have some additional groups such as ML-KEM. I.e., just keep the equivalency check.

File: test/sslapitest.c [hunk 3]
```diff
@@ -9782,22 +9782,30 @@ static int test_configuration_of_groups(void)
...
         || !TEST_size_t_eq(ctx->ext.supportedgroups_len, default_groups_len))
         goto end;
 
+#if (!defined(OPENSSL_NO_EC) || !defined(OPENSSL_NO_DH))
     if (!TEST_int_gt(SSL_CTX_set1_groups_list(ctx, "DEFAULT:-?P-256"), 0)
-#ifndef OPENSSL_NO_EC
+#else
+    if (!TEST_int_eq(SSL_CTX_set1_groups_list(ctx, "DEFAULT:-?P-256"), 0)
+#endif
```
Comment (t8m) [inline test/sslapitest.c:9799]: Same here, just skip the check.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by fwh-dc** [comment]
@t8m Thanks! Fixed now.

**Review by mattcaswell — APPROVED** [review]

**Comment by tmshort** [comment]
Ciphers use `!` for removal, do we want to introduce `-` for removal? Shouldn't we be consistent in these lists?

**Comment by mattcaswell** [comment]
Hmm. Good point. Yeah that might be better.

**Comment by t8m** [comment]
> Ciphers use `!` for removal, do we want to introduce `-` for removal? Shouldn't we be consistent in these lists?

Ciphers use both `!` and `-` for removal with `!` having the special meaning of _remove forever with no means to re-add_. So this doesn't introduce anything new and IMO `-` is better as the removal does not have any special meaning here.

**Comment by fwh-dc** [comment]
Yes I was trying to be consistent with:

https://github.com/openssl/openssl/blob/59f5f6c73cd2e1e2bd8ef405fdb6fadf0711f639/doc/man1/openssl-ciphers.pod.in#L136C1-L147C73

**Review by t8m — APPROVED** [review]

**Comment by tmshort** [comment]
OK, I've seen examples of ciphers with `!` in it, but I'm not sure I've seen an example with `-` (maybe it's there?).

**Comment by t8m** [comment]
From the openssl-ciphers(1) manpage:

       Each cipher string can be optionally preceded by the characters !, - or
       +.

       If  !  is  used then the ciphers are permanently deleted from the list.
       The ciphers deleted can never reappear in the list  even  if  they  are
       explicitly stated.

       If  -  is  used then the ciphers are deleted from the list, but some or
       all of the ciphers can be added again by later options.

       If + is used then the ciphers are moved to the end of  the  list.  This
       option  doesn't  add  any  new  ciphers it just moves matching existing
       ones.

**Comment by t8m** [comment]
Merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**